feat: Add tool_call support with json_schema for LLM

sarattha opened the pull request:
This pull request introduces several new functionalities and improvements to the `tensorrt_llm` package, focusing on function call detection and validation, as well as expanding response formats and cleaning up the codebase. The most important changes include adding utilities for detecting and parsing function calls, enhancing the response format options, and refactoring the code for better readability and maintainability.

### New Functionality:
* [`tensorrt_llm/serve/function_calling_utils.py`](diffhunk://#diff-f3462c2883e68f40e4ecafcbca3e59973c2c5518bac4f3951081ccabbd26b701R1-R77): Added utilities for detecting and parsing function calls from text generated by the model, including validating function arguments against JSON schema.

### Enhancements to Response Formats:
* [`tensorrt_llm/serve/openai_protocol.py`](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL48-R57): Expanded the `ResponseFormat` class to support a new type `json_schema`, and allowed extra fields in JSON schema responses.
* [`tensorrt_llm/serve/openai_protocol.py`](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL483-R508): Updated the `ChatCompletionRequest` class to support additional response formats and added new fields for better control over the response generation. [[1]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL483-R508) [[2]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL499-R543)

### Code Refactoring:
* [`tensorrt_llm/serve/openai_protocol.py`](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL79-R116): Refactored multiple classes and methods to improve code readability, including reformatting long lines, updating method signatures, and removing redundant code. [[1]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL79-R116) [[2]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL127-R139) [[3]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL241-R263) [[4]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL259-R288) [[5]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL305-R322) [[6]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL536) [[7]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL554) [[8]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL567-R607)

### Bug Fixes:
* [`tensorrt_llm/serve/openai_protocol.py`](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL318-R332): Fixed issues related to the handling of special tokens and stream options in the `ChatCompletionRequest` class. [[1]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL318-R332) [[2]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL444-R460) [[3]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL456-R473)

### Imports and Formatting:
* [`tensorrt_llm/serve/openai_protocol.py`](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL8-R13): Improved import statements and restructured code for better clarity and maintainability. [[1]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL8-R13) [[2]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL347-R367) [[3]](diffhunk://#diff-4b6ece0a77c7a186737d731752536269c760e4aa3e9ebc896f8023a37280d10aL367-R384)


Signed-off-by: Sarattha <j.sarattha@gmail.com>

sarattha committed:
Add tool_call support with json_schema for LLM
Signed-off-by: Sarattha <j.sarattha@gmail.com>

Signed-off-by: sarattha <j.sarattha@gmail.com>

ZhanruiSunCh committed:
infra: [TRTLLM-4370] Fix the build error when build GH200 image (#3229)

* infra: Fix the build error when build GH200 image

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* remove and update checkoutSource function

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

---------

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

ZhanruiSunCh committed:
infra: [TRTLLM-4308] Add Bot help (#3192)

* Add bot command help and check bot command

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* Fix permission error

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* Fix add comment

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* Fix

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* Fix review

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* Update bot-command.yml

Signed-off-by: Zhanrui Sun <184402041+ZhanruiSunCh@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Yanchao Lu <yanchaol@nvidia.com>
Signed-off-by: Zhanrui Sun <184402041+ZhanruiSunCh@users.noreply.github.com>

* Update .github/workflows/bot-command.yml

Co-authored-by: Yanchao Lu <yanchaol@nvidia.com>
Signed-off-by: Zhanrui Sun <184402041+ZhanruiSunCh@users.noreply.github.com>

* Fix pre-commit

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

---------

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>
Signed-off-by: Zhanrui Sun <184402041+ZhanruiSunCh@users.noreply.github.com>
Co-authored-by: Yanchao Lu <yanchaol@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

hyukn committed:
Fix minor issues in test_autotuner.py and loose the cache check for test gemms. (#3261)

This test can cause nondeterministic failures on CI with unexpected kernel profiling results. Given longer delay time or cache clear will not solve the issue. Thus, loose the test checks to avoid these false alarms.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

juney-nvidia commented:
@sarattha Thanks for the contribution.

@syuoni  @LinPoly @kaiyux can you help review this MR?

Thanks

June

yechank-nvidia committed:
doc: add supported-models on PyTorch example (#3179)

* doc: add supported-models on PyTorch example

Signed-off-by: yechank <161688079+yechank-nvidia@users.noreply.github.com>

* remove vision support from Llama3.2

Signed-off-by: yechank <161688079+yechank-nvidia@users.noreply.github.com>

---------

Signed-off-by: yechank <161688079+yechank-nvidia@users.noreply.github.com>
Co-authored-by: juney-nvidia <143764042+juney-nvidia@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

lfr-0531 committed:
fix deepseek-v3 mtp doc. (#3272)

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>
Co-authored-by: juney-nvidia <143764042+juney-nvidia@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

ZhanruiSunCh committed:
Fix bot check error when triggered by pull request (#3268)

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

kaiyux committed:
doc: Add serving section for DS V3 document (#3262)

Signed-off-by: Kaiyu Xie <26294424+kaiyux@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

yibinl-nvidia committed:
update FP4 quantize layout (#3045)

Signed-off-by: Yibin Li <109242046+yibinl-nvidia@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

pcastonguay committed:
chore: Refactor disaggregated serving scripts (#3073)

* chore: Refactor to reduce duplicated code in disagg server, reuse trtllm-serve

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

* Updating README, removing launch script

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

* Fixing integration tests

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

* Adding scripts to populate urls section of disagg config based on SLURM env vars

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

---------

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

Tracin committed:
AWQ support Modelopt ckpts. (#3258)

Signed-off-by: Tracin <10434017+Tracin@users.noreply.github.com>
Co-authored-by: QI JUN <22017000+QiJune@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

LinPoly committed:
doc: refactor trtllm-serve examples and doc (#3187)

Signed-off-by: Pengyun Lin <81065165+LinPoly@users.noreply.github.com>
Signed-off-by: Kaiyu Xie <26294424+kaiyux@users.noreply.github.com>
Co-authored-by: Kaiyu Xie <26294424+kaiyux@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

shaharmor98 committed:
feat: Support PeftCacheManager in Torch (#3186)

* Add PeftCacheManager implementation

Signed-off-by: Shahar Mor <smor@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

yuanjings-nvda committed:
fix vila test (#3042)

Signed-off-by: Yuanjing Shi <yuanjings@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

tburt-nv committed:
update internal_cutlass version.txt to d03df7b27 (#3279)

Signed-off-by: Tyler Burt <195370667+tburt-nv@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

QiJune committed:
fix deepseek multi gpu tests timeout (#3285)

Signed-off-by: junq <22017000+QiJune@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

Funatiq committed:
Reapply "refactor: Replace DecoderFinishedEvent with CudaEvent in decoder clas…" (#3183) (#3195)

* Reapply "refactor: Replace DecoderFinishedEvent with CudaEvent in decoder clas…" (#3183)

This reverts commit 75495730bca7a4c94feeed29b17014594f1a1598.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

* fixup! Reapply "refactor: Replace DecoderFinishedEvent with CudaEvent in decoder clas…" (#3183)

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

---------

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

jinyangyuan-nvidia committed:
perf: Use pinned H2D to reduce bubbles (#3147)

Signed-off-by: Jinyang Yuan <154768711+jinyangyuan-nvidia@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

qixiang-99 committed:
feat: no-cache attention in PyTorch workflow (#3085)

* init trtllm attn no cache

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* fix: fix the seq_len issue and attn metadata prepare for qwen reward model test

fix: fix minor bugs after rebase
Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* refactor: remove unnecessary debug logs and clean up commented code

refactor: update max_seq_len documentation and remove max_seq_len for decoder model contructor in PyTorchModelEngine
Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* refactor: update calculate_ref_result function to accept tensor inputs and mask type, enhance test_attention_no_cache to support FULL and CAUSAL masks

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* refactor: remove unused BERT attention metadata conversion method and add type assertion for no cache attention in PyTorchModelEngine

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* refactor: remove use_kv_cache parameter from attention function and related classes, update documentation for KV cache handling

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* refactor: implement setAttentionMaskType method for better mask type handling and remove unused conversion function

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* refactor: streamline KV cache handling by replacing direct member access with useKVCache method and simplify token per block assignment

remove Debug code.

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* refactor: Resolve comments for Python code

Simplify no cache attention metadata preparation and streamline related attributes in TrtllmAttentionMetadata

Removed the private method for converting to no cache attention metadata and integrated its logic into the prepare method. Updated the test for BERT sequence classification to reflect these changes and ensure proper handling of attention metadata.

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* docs: Add is_dummy_attention field to attention metadata for simulation operations

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* refactor: add KVCacheParams to attention backend interface and import relevant metadata classes

Updated the attention backend interface to include KVCacheParams and imported TrtllmAttentionMetadata and VanillaAttentionMetadata in model_engine.py for enhanced functionality.

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* fix: fix rebase format issue

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* fix: extend attention mask type handling in MHARunnerFixedParams

Added support for additional attention mask types (BIDIRECTIONAL, BIDIRECTIONALGLM, BLOCKSPARSE) in the MHARunnerFixedParams structure to fix the mapping issue between ContextAttentionMaskType and AttentionMaskType

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

* fix: enhance attention mask type handling in TllmGenFmhaRunnerParams

Updated the setAttentionMaskType method to include a switch-case structure for better handling of attention mask types, ensuring proper mapping and error handling for invalid types.

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

---------

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

Funatiq committed:
refactor: Expose DecoderState via bindings and integrate in TRTLLMDecoder (#3139)

* refactor: Expose DecoderState via bindings and integrate in TRTLLMDecoder

- Introduced a new `DecoderState` class in the C++ bindings, encapsulating key functionalities for managing decoding state.
- Adjusted the Python `TRTLLMDecoder` to access properties from `decoder_state`, ensuring consistency and clarity in the decoding process.

These changes streamline the decoder's architecture and enhance maintainability.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

* chore: Remove unused new_tokens from DecoderState bindings

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

---------

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

FrankD412 committed:
perf: Add total token throughput metric. (#3212)

Signed-off-by: Frank Di Natale <3429989+FrankD412@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

Superjomn committed:
chore: refactor the LlmArgs with Pydantic and migrate remaining pybinding configs to python (#3025)

* make LlmArgs Pydantic

Signed-off-by: Superjomn <328693+Superjomn@users.noreply.github.com>

* amending doc

fix api_stability

fix tests

Signed-off-by: Superjomn <328693+Superjomn@users.noreply.github.com>

* restore yaml groups

refine StackTrace

singleton

clean tests

Signed-off-by: Superjomn <328693+Superjomn@users.noreply.github.com>

* fix trtllm-bench

fix pytorch

Signed-off-by: Superjomn <328693+Superjomn@users.noreply.github.com>

* fix serve distagg

Signed-off-by: Superjomn <328693+Superjomn@users.noreply.github.com>

Signed-off-by: Superjomn <328693+Superjomn@users.noreply.github.com>

* fix

Signed-off-by: Superjomn <328693+Superjomn@users.noreply.github.com>

---------

Signed-off-by: Superjomn <328693+Superjomn@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

tburt-nv committed:
chore: remove usernames from comments (#3291)

Signed-off-by: Tyler Burt <195370667+tburt-nv@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

ncomly-nvidia committed:
L4 added to readme (#3301)

* Add L4 chart

Signed-off-by: Nick Comly <85702008+ncomly-nvidia@users.noreply.github.com>

* Add L4 to readme

Signed-off-by: Nick Comly <85702008+ncomly-nvidia@users.noreply.github.com>

* Add files via upload

Signed-off-by: Nick Comly <85702008+ncomly-nvidia@users.noreply.github.com>

* Update README.md

Signed-off-by: Nick Comly <85702008+ncomly-nvidia@users.noreply.github.com>

* Add files via upload

Signed-off-by: Nick Comly <85702008+ncomly-nvidia@users.noreply.github.com>

* Add files via upload

Signed-off-by: Nick Comly <85702008+ncomly-nvidia@users.noreply.github.com>

---------

Signed-off-by: Nick Comly <85702008+ncomly-nvidia@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

chuangz0 committed:
ucx interface (#3306)

Signed-off-by: Chuang Zhu <111838961+chuangz0@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

brb-nv committed:
test: Waive non-Llama Eagle tests (#3309)

Signed-off-by: sarattha <j.sarattha@gmail.com>

lfr-0531 committed:
fix the py_decoding_iter update in decoder. (#3297)

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>
Co-authored-by: QI JUN <22017000+QiJune@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

Shunkangz committed:
fix: fix attentionDP padding request type (#3299)

* Fix attentionDP padding request type

Signed-off-by: Shunkang <182541032+Shunkangz@users.noreply.github.co>

* Refactor import

Signed-off-by: Shunkang <182541032+Shunkangz@users.noreply.github.co>

---------

Signed-off-by: Shunkang <182541032+Shunkangz@users.noreply.github.co>
Co-authored-by: Shunkang <182541032+Shunkangz@users.noreply.github.co>
Signed-off-by: sarattha <j.sarattha@gmail.com>

nv-guomingz committed:
chore: update internal cutlass library base #2981 and #3165. (#3308)

Signed-off-by: nv-guomingz <37257613+nv-guomingz@users.noreply.github.com>
Co-authored-by: nv-guomingz <37257613+nv-guomingz@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

QiJune committed:
chore: waive a timeout multi-GPU test case (#3310)

* debug CI timeout issue

Signed-off-by: junq <22017000+QiJune@users.noreply.github.com>

* fix

Signed-off-by: junq <22017000+QiJune@users.noreply.github.com>

* waive timeout case

Signed-off-by: junq <22017000+QiJune@users.noreply.github.com>

---------

Signed-off-by: junq <22017000+QiJune@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

bobboli committed:
feat: Add support for FP8 MLA on Hopper and Blackwell. (#3190)

* fp8 kv + bf16 ctx MLA + fp8 gen MLA

Use BF16 for context MLA.
mFP8GenerationMLA and mFP8ContextFMHA shouldn't be enabled together.

Allow mSM==90 for mFP8GenerationMLA==true.
For FMHA, dataTypeKv should be FP8.

For FP8 MLA generation, the output is still in BF16.

Refine debug info for FMHA kernel metadata.

Use inputType, outputType, SM together to hash kernel list.

Add FP8 MLA generation FMHA kernel.

Special WAR of NUM_COMPUTE_GROUPS for MLA generation kernel.

Separate the implementation of fused_multihead_attention_v2.h to CPP and print some debug info if checkIfKernelExist fails.

Refine debug info in fused_multihead_attention_v2.cpp

Correct FP8 MLA metadata.

New kernel provided by Yuxin, which outputs BF16.

smem size is not set correctly, which will lead to illegal mem access.

Yuxin fixed the error in FMHA MLA kernel: previously the BF16 isn't correctly written: some parts are repeatedly written, while some others are untouched.

There are two bmm1 scales that should be set correctly.

New kernel generated by Yuxin.

Modificatiosn to common/attentionOp for FP8 MLA on Hopper using FMHA.

Not necessary. If mFP8GenerationMLA, is_fp8_out is false, so mFP8ContextFMHA is false.

Skip a check in fmhaDispatcher.

Modifications in fmhaRunner:
- Debug dump.
- if (!isFP8GenerationMLA) skips a lot of flag setting.
- TMA descriptor modification for qo (by Yuxin).

Cleanup debug output.

Clean up o tma descriptor modifications.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* Resolve conflicts.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* Apply the patch of FP8 FlashMLA and resolve conflicts.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* Fix compilation error.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* Fix compile error.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* pick blackwell support

Signed-off-by: Dylan Chen <191843203+DylanChen-NV@users.noreply.github.com>

* Add copyright notice to fused_multihead_attention_v2.cpp.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* Add license.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* Add missing license.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* Exclude building flashMLA kernels under sm90.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* Revert "Exclude building flashMLA kernels under sm90."

    This reverts commit f0c859d4598b718994b8df7e41dd19eed9fd28fd.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

* Use macro to skip compiling FlashMLA for non sm90 targets.

Signed-off-by: Bo Li <bobboli0202@gmail.com>

---------

Signed-off-by: Bo Li <bobboli0202@gmail.com>
Signed-off-by: Dylan Chen <191843203+DylanChen-NV@users.noreply.github.com>
Co-authored-by: Dylan Chen <ziqingc@nvidia.com>
Co-authored-by: Dylan Chen <191843203+DylanChen-NV@users.noreply.github.com>
Co-authored-by: QI JUN <22017000+QiJune@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

yiqingy0 committed:
chore: Blossom debug hook (#3091)

Signed-off-by: Yiqing Yan <yiqingy@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

pansicheng committed:
feat: support abort disconnected requests (#3214)

Signed-off-by: pansicheng <sicheng.pan.chn@gmail.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

amitz-nv committed:
Fix torch nvsmall through pyexecutor and fix its TP support (#3238)

* Fix NemotronNAS support

Signed-off-by: Amit Zuker <203509407+amitz-nv@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

lowsfer committed:
Support speculative decoding with Hopper XQA (#3269)

Signed-off-by: Yao Yao <lowsfer@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

yweng0828 committed:
test: fix conflicting test names (#3316)

Signed-off-by: Yue Weng <25103990+yweng0828@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

lucifer1004 committed:
feat: use NVRTC for DeepGEMM JIT compilation (#3239)

* feat: use NVRTC for DeepGEMM JIT compilation

Signed-off-by: Zihua Wu

* fix: add license

Signed-off-by: Zihua Wu

* feat: store NVRTC JIT results in memory by default

Signed-off-by: Zihua Wu

* feat: refinement

Signed-off-by: Zihua Wu

* feat: refinement

Signed-off-by: Zihua Wu

* test: set timeout to 7200

Signed-off-by: Zihua Wu

---------

Signed-off-by: Zihua Wu
Signed-off-by: sarattha <j.sarattha@gmail.com>

chuangz0 committed:
update readme for disaggregated (#3323)

Signed-off-by: Chuang Zhu <111838961+chuangz0@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

syuoni committed:
test: Accuracy test improvement (Part 3.3): Move DeepSeek tests (#3260)

add skip

fix

fix

update

update test list

fixqa list

move bf16 to postmerge

Signed-off-by: Enwei Zhu <21126786+syuoni@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

1ytic committed:
fix: redrafter sampling (#3278)

* Fix redrafter sampling

Signed-off-by: Ivan Sorokin <isorokin@nvidia.com>

* Rename redrafter bream search var

Signed-off-by: Ivan Sorokin <isorokin@nvidia.com>

* Remove _beam_search_candidates_v0

Signed-off-by: Ivan Sorokin <isorokin@nvidia.com>

* Remove unused import

Signed-off-by: Ivan Sorokin <isorokin@nvidia.com>

---------

Signed-off-by: Ivan Sorokin <isorokin@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

yilin-void committed:
fix: runtime error in est_deepseek_allreduce.py (#3226)

Signed-off-by: Yilin Zhang <18275976+yilin-void@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

pcastonguay committed:
feat: Add option to run disaggregated serving without ctx servers,… (#3243)

* feat: Add option to run disaggregated serving without ctx servers, to benchmark gen only

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

* Fixing comment in sanity check

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

---------

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

kaiyux committed:
breaking change: perf: Make ipc_periodically the default responses_handler (#3102)

Signed-off-by: Kaiyu Xie <26294424+kaiyux@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

chuangz0 committed:
feat: use cudaMalloc to allocate kvCache (#3303)

Signed-off-by: sarattha <j.sarattha@gmail.com>

lucifer1004 committed:
fix: revert extra cmake var (#3351)

Signed-off-by: Zihua Wu <13583761+lucifer1004@users.noreply.github.com>
Co-authored-by: QI JUN <22017000+QiJune@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

MinaHuai committed:
add tp=2 ci test for vision encoder (#3319)

Signed-off-by: mhuai <mhuai@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

lfr-0531 committed:
Waive unittest/trt/model/test_mamba.py::TestMamba::test_loaders_mamba_130m_hf_from_checkpoint. Will fix it later. (#3356)

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

lucifer1004 committed:
feat: enable DeepGEMM by default (#3341)

Signed-off-by: Zihua Wu <13583761+lucifer1004@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

hyukn committed:
feat: Apply the new torch-flow compatible AutoTuner to both Fused MoE and NVFP4 Linear operators. (#3151)

* Several optimizations and fixings on the Autotuner.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

* Apply the new Python side Autotuner on current linear for nvFP4 data type.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

* Apply the new Python side Autotuner on MoE op
* Remove routers from cache key to improve inference perf
* Prevent unnecessary code profiling. Use do_preparation keyword to select which part should be executed during before evaluating any tactic.
* Remove try-catch inside moe profiling process.
* Move default tactic -1 to 0 transforms in cpp runner.
* Revise relavant tests.
* Predefined the bucketizing strategy for fused_moe

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

* Add specific_profile support for AutoTuner to bypass the standard cache search process for perf optimization
* Add specific_profile for moe
* Add specific profile for linear

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

* Fixing and revising according to reviewer's suggestions.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

* Use lru_cache for inference pref optimization.
* Revert gen_custom_cache_key feature

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

* Replace runner with runner id to achieve a serializable cache.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

* Code clean up and minor fixings.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

* Move all tunable runners and custom ops into torch_custom_ops.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

* Treat min_latency_mode as a independent dynamic tensor. Modify get_valid_tactics to suit for it.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

---------

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

LinPoly committed:
Use llm.tokenizer in OpenAIServer (#3199)

Signed-off-by: Pengyun Lin <81065165+LinPoly@users.noreply.github.com>
Co-authored-by: Yan Chunwei <328693+Superjomn@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

syuoni committed:
test: Accuracy test improvement (Part 3.4): Move LLaMA tests (#3350)

Signed-off-by: Enwei Zhu <21126786+syuoni@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

Superjomn committed:
clean up trtllm-llmapi-launch logs (#3358)

Signed-off-by: Superjomn <328693+Superjomn@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

amirkl94 committed:
fix: Fix p-tuning test bug (#3326)

* fix: Fix p-tuning test bug

* A change in the vocab_size calculation for T5Tokenizer,
introduced in transformers version 4.34, caused addition of incorrect vtokens for ptuning.
In general, instead of adding tokens which are outside the vocabulary, tokens inside the vocabulary were added.

Signed-off-by: Amir Klein <203507526+amirkl94@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

chuangz0 committed:
disagg test single h100 (#3353)

Signed-off-by: sarattha <j.sarattha@gmail.com>

ZhanruiSunCh committed:
infra: Fix bot help error when " in bot command (#3314)

* Fix bot help error when " in bot command

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* Delete a.txt

Signed-off-by: Yanchao Lu <yanchaol@nvidia.com>

---------

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>
Signed-off-by: Yanchao Lu <yanchaol@nvidia.com>
Co-authored-by: Yanchao Lu <yanchaol@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

liji-nv committed:
feat: Introduce UB allocator for pytorch flow (#3257)

* Instead of allocating UserBuffers at beginning of runtime, UB buffers
  are now managed with global allocator. The allocator will dynamically
assign free UB buffer or allocate new buffer for torch tensor. It makes
userbuffers easier to use.

* In common usecase, the Userbuffers will be allocated correctly during
  warm up stage. There is no dynamic allocation during inference.

* UB fusion pattern is rewroten using the new UB Allocator. It contains
  following passes:

1. Fuse Quant with allreduce, replace with UB impl, and insert a
   copy_to_userbuffers. Currently the normal allreduce still does not
   support FP8 quant. So this need to be done in UB pass
2. Convert all supported allreduce with UB and insert copy_to_userbuffers.
3. Fuse op before ar with the copy_to_userbuffers. So the op directly
   writes to the userbuffer
4. Remove userbuffers finalize if the output is connect to another UB
   allreduce.

Signed-off-by: Jin Li <59594262+liji-nv@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

yuxianq committed:
Add thread leak check and fix thread/memory leak issues. (#3270)

Signed-off-by: Yuxian Qiu <142763828+yuxianq@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

yilin-void committed:
feat: fix and improve allreduce and fusion kernels (#3064)

Signed-off-by: Yilin Zhang <18275976+yilin-void@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

ZhanruiSunCh committed:
chore: bump version to 0.19.0.dev2025041500 (#3360)

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

pcastonguay committed:
chore: Adding DS V3-lite tests with overlap + cuda graph (#3342)

* chore: Adding DS V3-lite tests with overlap + cuda graph

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

* Fixing pre-commit

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

---------

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

sugunav14 committed:
feat: [TRTLLM-3510] DeepseekV3 support in AutoDeploy (#3281)

Signed-off-by: Suguna Velury <178320438+sugunav14@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

kaiyux commented on tensorrt_llm/serve/openai_protocol.py:
```diff
@@ -227,8 +239,6 @@ def to_sampling_params(self) -> SamplingParams:
             skip_special_tokens=self.skip_special_tokens,
             spaces_between_special_tokens=self.spaces_between_special_tokens,
             truncate_prompt_tokens=self.truncate_prompt_tokens,
-            return_context_logits=self.return_context_logits,
```
Why is this line removed?

sarattha replied:
My mistake, revert back to the original.

kaiyux commented on tensorrt_llm/serve/openai_protocol.py:
```diff
@@ -564,28 +585,26 @@ def model_post_init(self, __context: Any) -> None:
-                raise ValueError(
-                    "When using `tool_choice`, `tools` must be set.")
-        return data
+    # @model_validator(mode="before")
```
Please remove the function instead of commenting it.

sarattha replied:
Done.

kaiyux commented on tensorrt_llm/serve/postprocess_handlers.py:
```diff
@@ -158,36 +171,47 @@ def yield_first_chat(num_tokens: int,
+        # Handle explicit function call
+        if args.tool_choice and isinstance(args.tool_choice, ChatCompletionNamedToolChoiceParam):
+            # Handle explicit function call (existing code)
+            pass
```
Seem to miss some implementation?

sarattha replied:
Yes, I intentionally left the implementation incomplete because I wasn’t certain about the optimal way to handle explicit function calls at that time. I used a placeholder (pass) to indicate where a proper implementation should go and wanted to open the discussion for suggestions—perhaps something along the lines of an explicit tool call like those in the OpenAI API (e.g., RAG or the Web search tool). If this functionality turns out to be unnecessary, I’m happy to remove it. For now, I’ve included my proposed implementation for explicit tool calls in this commit and look forward to your feedback.

kaiyux reviewed (commented):
Thanks for the contribution!

It seems that a lot of lines are changed only on format, can you please try format the code following the guidance [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CONTRIBUTING.md#coding-style)? 

Besides, it seems to me some implementation was removed by `pass`, is that expected? Please correct me if my understanding is incorrect, thanks!

syuoni commented:
HI @sarattha ,

Thanks for the PR! Is there a usage example for this feature?

wili-65535 committed:
Feat: Variable-Beam-Width-Search (VBWS) part3 (#3338)

* feat/Variable-Beam-Width-Search-Part3, v1.0

Signed-off-by: wili-65535 <wili-65535@user.noreply.github.com>

* feat/Variable-Beam-Width-Search-Part3, v1.1

Signed-off-by: wili-65535 <wili-65535@user.noreply.github.com>

* feat/Variable-Beam-Width-Search-Part3, v1.2

Signed-off-by: wili-65535 <wili-65535@user.noreply.github.com>

---------

Signed-off-by: wili-65535 <wili-65535@user.noreply.github.com>
Co-authored-by: wili-65535 <wili-65535@user.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

ZhanruiSunCh committed:
infra: [TRTLLM-4450] Support more files for pytorch only mode (#3365)

* infra: [TRTLLM-4450] Support more files for pytorch only mode

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* Test pytorch only mode

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* Revert "Test pytorch only mode"

This reverts commit b32f54d7858bd2432251734bc7b31669147ed94b.

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>

* Fix review

Signed-off-by: Zhanrui Sun <184402041+ZhanruiSunCh@users.noreply.github.com>

---------

Signed-off-by: ZhanruiSunCh <184402041+ZhanruiSunCh@users.noreply.github.com>
Signed-off-by: Zhanrui Sun <184402041+ZhanruiSunCh@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

yuxianq committed:
chore: Refine attention backend interface. (#3271)

Refine attention backend interface.

Signed-off-by: Yuxian Qiu <142763828+yuxianq@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

mikeiovine committed:
Add Llama 4 (#3302)

Signed-off-by: Mike Iovine <miovine@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

Tracin committed:
Fix miss bias add for FP4Linear. (#3361)

Signed-off-by: Tracin <10434017+Tracin@users.noreply.github.com>
Co-authored-by: QI JUN <22017000+QiJune@users.noreply.github.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

Tabrizian committed:
test: Add single gpu disaggregated tests (#3295)

* test: Add single gpu disaggregated tests

Signed-off-by: Iman Tabrizian <itabrizian@nvidia.com>

* Add deepseek with overlap tests

Signed-off-by: Iman Tabrizian <itabrizian@nvidia.com>

* Use updated prompt

Signed-off-by: Iman Tabrizian <itabrizian@nvidia.com>

* Move test to disaggregated folder

Signed-off-by: Iman Tabrizian <itabrizian@nvidia.com>

---------

Signed-off-by: Iman Tabrizian <itabrizian@nvidia.com>
Signed-off-by: sarattha <j.sarattha@gmail.com>

sarattha committed:
Remove commented function, revert back deleted line and implement explicit function call

Signed-off-by: sarattha <j.sarattha@gmail.com>

sarattha committed:
Apply proper black format

Signed-off-by: sarattha <j.sarattha@gmail.com>

sarattha committed:
Fix wrong formatted line length

Signed-off-by: sarattha <j.sarattha@gmail.com>

sarattha commented:
> HI @sarattha ,
> 
> Thanks for the PR! Is there a usage example for this feature?

Absolutely. I'm currently working on a project that deploys a multi-agent system using the OpenAI Agent SDK framework with local LLM models (e.g., Llama3.3 and Qwen). In our framework, an agent is able to call tools to execute more complex tasks. The mechanism involves embedding a `<tool_call></tool_call>` tag within the generated message along with a `json_schema` response type.

Since I also use TensorRT-LLM as the inference engine for our LLM models, it's essential that the TensorRT-LLM serving handles tool calls in a manner similar to the OpenAI API.

sarattha committed:
fix: revert missing original codes

Signed-off-by: sarattha <j.sarattha@gmail.com>

sarattha commented:
Sorry I mess up this PR. I will create new one.
